source/common/upstream/upstream_impl.cc

```diff
@@ -1432,16 +1430,18 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
 
   // Remove hosts from current_priority_hosts that were matched to an existing host in the previous
   // loop.
-  for (auto itr = current_priority_hosts.begin(); itr != current_priority_hosts.end();) {
-    auto existing_itr = existing_hosts_for_current_priority.find((*itr)->address()->asString());
+  auto current_erase_from = current_priority_hosts.end();
```

Comment:
I don't want to refactor this whole function in this PR but it's quite long. I think there's an opportunity to simplify it and also improve unit testing of the new algorithm by encapsulating this logic into a helper method in source/common/common/utility.y:
```
template<class Container, class T>
void Utility::removeMatchingElements(Container&, std::function<bool(const T&)> predicate);
```
This has 3 benefits:
 - marginally making this large function more readable
 - enabling you to do targeted unit-testing of the new algorithm
 - enabling easier re-use of this pattern elsewhere should it be needed.

Actually I think from the usage 
WDYT?

Reply:
I think this is reasonable idea but I'd prefer to separate functional & behavioral changes from refactoring in the interest of purity and ease of testing / review. I'm happy to take this on in a followup, and I acknowledge I've introduced some duplicated, unattractive code in this PR.

Reply:
well the reason I'd argue it's in-scope for this PR is that you introduced a new algorithm into the middle of an existing algorithm, but we can't unit-test the algorithm in situ.

I think this should be pretty simple no?

Reply:
Touché, can't hurt to give it a try.

Reply:
I'm not confident I've ordered utility.h correctly.